Native ATen/Parallel backend

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20157 Support for Eigen thread pool
* #20480 Native TBB parallel backend
* #20454 Restore TBB module
* **#20087 Native ATen/Parallel backend**

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

Differential Revision: [D15248663](https://our.internmc.facebook.com/intern/diff/D15248663)

Commit:
Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

Commit:
Update on "[wip] Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

dzhulgakov commented on aten/src/ATen/PTThreadPool.h:
```diff
@@ -4,6 +4,16 @@
 
+// Checks external number of threads setting and returns
+// default value (hardware_concurrency()) if it's negative
+inline size_t check_and_get_pool_size(int nthreads) {
```
does it need to be an interface function? feels sketchy

ilia-cher replied:
I need to put the 'default thread pool size' somewhere, can move it

dzhulgakov commented on aten/src/ATen/Parallel_native.cpp:
```diff
@@ -0,0 +1,112 @@
+}
+
+void set_num_threads(size_t nthreads) {
+  if (nthreads == 0) {
```
<= ?

ilia-cher replied:
nthreads is size_t

dzhulgakov commented on aten/src/ATen/Parallel_native.cpp:
```diff
@@ -0,0 +1,112 @@
+  }
+}
+
+size_t get_num_threads() {
```
this function is called just next to actual computation. Why can't it be just _get_intraop_pool().size()?

ilia-cher replied:
I don't want to create a thread pool without need, because after this I can't change its size

dzhulgakov commented on aten/src/ATen/Parallel_native.cpp:
```diff
@@ -0,0 +1,112 @@
+  }
+}
+
+size_t get_num_threads() {
```
I find it a bit confusing that it returns pool+1. I guess it's ok but I'd document it in .h file

ilia-cher replied:
yep, I think it's better to treat user setting N as 1 master + (N-1) in thread pool, going to upd

dzhulgakov commented on aten/src/ATen/Parallel_native.h:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    int64_t first_task_end = std::min(end, (int64_t)(chunk_size + begin));
+    internal::_set_in_parallel_region(true);
```
don't you need to set it also on child threads? (i thought that's omp's semantics)

in that case you can also just use get_thread_num

ilia-cher replied:
we do return true for child threads because we call `internal::_get_intraop_pool().inThreadPool` - but i'll also set `internal::_set_in_parallel_region(true)` for them, then we probably don't need `internal::_get_intraop_pool().inThreadPool`

dzhulgakov commented on aten/src/ATen/Parallel_native.cpp:
```diff
@@ -0,0 +1,112 @@
+  }
+}
+
+int get_thread_num() {
```
this function is fishy. It seems to mix logically-set thread id (which is property of the parallel primitive you're executing) and a global thread id for the pool. I think they are different and you likely need only the former one.

ilia-cher replied:
fair, I originally wrote it without thread local (and that was a bug because thread numbers in OMP are logical task ids), so I'll just remove the internal::_get_intraop_pool().threadNum() usage

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

ilia-cher commented:
Addressed comments

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

jamesr66a commented on aten/src/ATen/Parallel_native.h:
```diff
@@ -0,0 +1,181 @@
+CAFFE2_API void _unset_thread_num();
+}
+
+template <class F>
```
Can you add comments describing what these do at a high level? If these are to be used in operator implementations, we should have some really good docs so people know what to do e.g. for setting `grain_size`

ilia-cher replied:
Will add, I'd suggest only to add this into generic interface file (Parallel.h) - we already have a good comment on `parallel_reduce`, need to add some for `parallel_for` there too

jamesr66a commented on aten/src/ATen/Parallel.h:
```diff
@@ -97,3 +97,7 @@ CAFFE2_API void launch(const std::function<void()>& func);
 #if AT_PARALLEL_OPENMP
```
Are `AT_PARALLEL_OPENMP` and `AT_PARALLEL_NATIVE` required to be mutually exclusive?

ilia-cher replied:
Right, they are mutually exclusive

jamesr66a reviewed (commented):
Pretty exciting! Small comments

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update base for Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

cpuhrsch commented on aten/src/ATen/ParallelNative.h:
```diff
@@ -0,0 +1,154 @@
+    chunk_size = std::max((size_t)grain_size, chunk_size);
+    size_t num_tasks = divup((end - begin), chunk_size);
+
+    auto task = [&](int64_t task_id, int64_t local_start, int64_t local_end) {
```
Is the capture actually necessary here? Looks like it doesn't use anything from the outer scope.

ilia-cher replied:
will upd

cpuhrsch commented on aten/src/ATen/ParallelNative.h:
```diff
@@ -0,0 +1,154 @@
+      futures[task_id]->wait();
+    }
+
+    return std::accumulate(
```
std::accumulate again - see https://github.com/pytorch/pytorch/pull/20057#discussion_r284838795

ilia-cher replied:
will replace with a loop

cpuhrsch commented on setup.py:
```diff
@@ -154,6 +154,7 @@
 #     parallel backend to use for intra- and inter-op parallelism
 #     possible values:
 #       OPENMP - use OpenMP for intra-op and native backend for inter-op tasks
+#       NATIVE - use native thread pool for both intra- and inter-op tasks
```
Do we eventually want to distinguish between inter and intra or have something like OPENMP_NATIVE and NATIVE_NATIVE to show which one is used for which?

ilia-cher replied:
My thinking is that we should have:
- OPENMP - OpenMP for intra-op, native thread pool for inter-op
- NATIVE - using native thread pool for both intra- and inter- (actually two separate pools)
- NATIVE_TBB - (next in stack) tbb for intra-op, native thread pool for inter-op
- potentially, just TBB - TBB for both intra- and inter-

cpuhrsch commented on aten/src/ATen/ParallelNative.h:
```diff
@@ -0,0 +1,154 @@
+        int64_t local_end = std::min(end, (int64_t)(chunk_size + local_start));
+        internal::_get_intraop_pool().run(
+            // copy future_ptr, task_id, local_start, local_end
+            [&, future_ptr, task_id, local_start, local_end]() {
```
Does this need the & capture? I'm mostly worried about pulling in objects that aren't needed but could cause performance implications.

ilia-cher replied:
will upd

cpuhrsch reviewed (approved):
Assuming comments get resolved this looks nice!

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

dzhulgakov commented on aten/src/ATen/ParallelNative.cpp:
```diff
@@ -0,0 +1,120 @@
+  } else if (nthreads == NOT_SET) {
+    return TaskThreadPoolBase::defaultNumThreads();
+  } else {
+    TORCH_CHECK(nthreads == CONSUMED);
```
TORCH_INTERNAL_ASSERT?

ilia-cher replied:
makes sense, will upd

dzhulgakov reviewed (approved):
Looks good!

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head

Commit:
Update on "Native ATen/Parallel backend"

Native ATen/Parallel backend

Summary:
Implementing ATen/Parallel withtout OpenMP using native thread pool

Test Plan:
PARALLEL_BACKEND=NATIVE BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20087 gh/ilia-cher/15/head
